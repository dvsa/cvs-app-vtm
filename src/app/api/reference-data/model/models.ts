export * from './reference-data.model';
