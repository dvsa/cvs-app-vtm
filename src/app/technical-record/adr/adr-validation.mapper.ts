import { Observable, Subject } from 'rxjs';
import { Injectable } from '@angular/core';

enum CHECKER {
  BATTERY = 'battery',
  TANK = 'tank',
  NONE = 'na'
}

export enum STATUS {
  MANDATORY = 'MANDATORY',
  HIDDEN = 'HIDDEN'
}

export interface ValidationState {
  tankDetailsEdit?: STATUS;
  tankDocuments?: STATUS;
  tankInspectionsEdit?: STATUS;
  memoEdit?: STATUS;
  batteryListApplicableEdit?: STATUS;
}

@Injectable({ providedIn: 'root' })
export class ValidationMapper {
  state = new Subject<ValidationState>();

  constructor() {}

  mapVehicleTypeToValidationState(selectedType: string): void {
    const validationStatus = {
      [CHECKER.BATTERY]: () => this.getBatteryTypeValidationStatus(),
      [CHECKER.TANK]: () => this.getTankTypeValidationStatus(),
      [CHECKER.NONE]: () => this.getOtherTypeValidationStatus()
    };

    const selectedState: ValidationState = validationStatus[this.getType(selectedType)]();
    this.state.next(selectedState);
  }

  getVehicleTypeState(): Observable<ValidationState> {
    return this.state.asObservable();
  }

  private getType(type: string): string {
    if (type.trim().includes(CHECKER.BATTERY)) {
      return CHECKER.BATTERY;
    }

    if (type.includes(CHECKER.TANK)) {
      return CHECKER.TANK;
    }

    return CHECKER.NONE;
  }

  private getBatteryTypeValidationStatus(): ValidationState {
    return {
      tankDetailsEdit: STATUS.MANDATORY
    };
  }

  private getTankTypeValidationStatus(): ValidationState {
    return {
      tankDetailsEdit: STATUS.MANDATORY,
      batteryListApplicableEdit: STATUS.HIDDEN
    };
  }

  private getOtherTypeValidationStatus(): ValidationState {
    return {
      tankDetailsEdit: STATUS.HIDDEN,
      tankDocuments: STATUS.HIDDEN,
      tankInspectionsEdit: STATUS.HIDDEN,
      memoEdit: STATUS.HIDDEN,
      batteryListApplicableEdit: STATUS.HIDDEN
    };
  }
}
