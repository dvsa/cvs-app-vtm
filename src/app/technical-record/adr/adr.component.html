<dl class="govuk-summary-list">
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key">
      Able to carry dangerous goods
      <ng-container *ngIf="!!editState">
        <div id="radioADR" class="govuk-radios govuk-radios--inline govuk-!-margin-top-4">
          <div
            class="govuk-radios__item"
            *ngFor="let option of booleanOptions | keyvalue: unsorted"
          >
            <input
              class="govuk-radios__input"
              id="test-edit-hasADR-{{ option.key }}"
              name="hasADR"
              type="radio"
              [checked]="showAdrView === option.value"
              [value]="option.value"
              (change)="switchAdrDisplay($event)"
            />
            <label class="govuk-label govuk-radios__label" for="hasADR-{{ option.key }}">
              {{ option.key }}
            </label>
          </div>
        </div>
      </ng-container>
    </dt>

    <dd id="test-ableToCarry" class="govuk-summary-list__value">
      <div *ngIf="!editState">
        <span> {{ showAdrView ? 'Yes' : 'No' }}</span>
      </div>
    </dd>
    <dd class="govuk-summary-list__value"></dd>
  </div>
</dl>

<ng-container *ngIf="showAdrView">
  <!-- Applicant Details -->
  <vtm-applicant-details [edit]="!!editState" [adrDetails]="adrDetails"></vtm-applicant-details>

  <!-- ADR Details -->
  <vtm-adr-details
    [edit]="!!editState"
    [adrDetails]="adrDetails"
    [metaData]="vehicleMetaData"
  ></vtm-adr-details>

  <!-- Tank Details -->
  <vtm-tank-details [edit]="!!editState" [adrDetails]="adrDetails"></vtm-tank-details>

  <!-- Tank Inspections -->
  <vtm-tank-inpections [edit]="!!editState" [adrDetails]="adrDetails"></vtm-tank-inpections>

  <!-- Memo -->
  <vtm-memo [edit]="!!editState" [adrDetails]="adrDetails"></vtm-memo>

  <!-- Tank Documents -->
  <vtm-tank-documents [edit]="!!editState" [adrDetails]="adrDetails"></vtm-tank-documents>

  <!-- Battery List Applicable -->
  <vtm-battery-list-applicable
    [edit]="!!editState"
    [adrDetails]="adrDetails"
  ></vtm-battery-list-applicable>

  <!-- Declaration Seen -->
  <vtm-declaration-seen [edit]="!!editState" [adrDetails]="adrDetails"></vtm-declaration-seen>

  <!-- Certificate -->
  <vtm-certificate [edit]="!!editState" [adrDetails]="adrDetails"></vtm-certificate>

  <vtm-additional-adr-details
    [edit]="!!editState"
    [adrDetails]="adrDetails"
  ></vtm-additional-adr-details>
</ng-container>
