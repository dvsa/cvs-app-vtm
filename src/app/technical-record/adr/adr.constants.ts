const EXPLOSIVES2 = 'Explosives (type 2)';
const EXPLOSIVES3 = 'Explosives (type 3)';

export const EXPLOSIVE_TYPES = [EXPLOSIVES2, EXPLOSIVES3];

export const STATEMENT = 'statement';
export const PRODUCT_LIST = 'productList';

export const UUID_REGEX = /[0-9a-fA-F]{8}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{12}/g;
