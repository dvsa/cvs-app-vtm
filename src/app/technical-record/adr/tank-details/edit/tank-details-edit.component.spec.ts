import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import {
  ReactiveFormsModule,
  FormsModule,
  FormGroupDirective,
  FormGroup,
  AbstractControl
} from '@angular/forms';
import { By } from '@angular/platform-browser';
import { of } from 'rxjs';

import { TESTING_UTILS } from '@app/utils/testing.utils';
import { SharedModule } from '@app/shared';
import { TankDetailsEditComponent } from './tank-details-edit.component';
import { TankStatement } from '@app/models/Tank';
import { SUBSTANCES } from '@app/app.enums';
import { STATEMENT, PRODUCT_LIST } from '../../adr.constants';
import { ValidationMapper, STATUS } from '../../adr-validation.mapper';

describe('TankDetailsEditComponent', () => {
  let component: TankDetailsEditComponent;
  let fixture: ComponentFixture<TankDetailsEditComponent>;
  let validationMapper: ValidationMapper;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [FormsModule, ReactiveFormsModule, SharedModule],
      declarations: [TankDetailsEditComponent],
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
      providers: [
        ValidationMapper,
        FormGroupDirective,
        {
          provide: FormGroupDirective,
          useValue: TESTING_UTILS.mockFormGroupDirective({
            techRecord: new FormGroup({
              adrDetails: new FormGroup({})
            }) as AbstractControl
          })
        }
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    validationMapper = TestBed.get(ValidationMapper);
    fixture = TestBed.createComponent(TankDetailsEditComponent);
    component = fixture.componentInstance;
    component.tank = TESTING_UTILS.mockTank();
  });

  it('should create with initialized form controls', () => {
    fixture.detectChanges();

    expect(component).toBeDefined();
    expect(fixture).toMatchSnapshot();
  });

  it('should create with multiple UN number controls', () => {
    component.tank = TESTING_UTILS.mockTank({
      tankStatement: { productListUnNo: ['listUN1', 'listUN2'] } as TankStatement
    });
    fixture.detectChanges();

    expect(component.productListUnNo.controls.length).toBe(2);
    expect(fixture).toMatchSnapshot();
  });

  it('should call vehicleTypeSelected on validationMapper if vehicle type exist', () => {
    const type = TESTING_UTILS.mockVehicleDetails().type;
    spyOn(validationMapper, 'vehicleTypeSelected');
    component.vehicleType = type;

    fixture.detectChanges();

    expect(validationMapper.vehicleTypeSelected).toHaveBeenCalledWith(type);
  });

  it('should add a UN number control to the one created by default', () => {
    fixture.detectChanges();
    component.addUNNumberControl();

    expect(component.productListUnNo.controls.length).toBe(2);
  });

  it('should remove all UN number controls but one', () => {
    component.tank = TESTING_UTILS.mockTank({
      tankStatement: { productListUnNo: ['listUN1', 'listUN2'] } as TankStatement
    });
    fixture.detectChanges();
    component.removeUNNumberControlButOne();

    expect(component.productListUnNo.controls.length).toBe(1);
  });

  describe('handleFormChanges', () => {
    it('should reset controls when substance permitted is selected', () => {
      fixture.detectChanges();

      const selectedSubstance: HTMLElement = fixture.debugElement.query(
        By.css('[id=test-edit-substancesPermitted-permitted]')
      ).nativeElement;
      selectedSubstance.click();

      expect(component.substancesPermitted.value).toEqual(SUBSTANCES.PERMITTED);
      expect(component.substanceReferenceSelect.value).toBeNull();
      expect(component.tankStatement.get('statement').value).toBeNull();
      expect(component.tankStatement.get('productListRefNo').value).toBeNull();
      expect(component.productListUnNo.value).toEqual([null]);
      expect(component.tankStatement.get('productList').value).toBeNull();
    });

    it('should reset controls within product list when Statement option is selected', () => {
      fixture.detectChanges();

      component.substanceReferenceSelect.setValue(STATEMENT);
      expect(component.tankStatement.get('productListRefNo').value).toBeNull();
      expect(component.productListUnNo.value).toEqual([null]);
      expect(component.tankStatement.get('productList').value).toBeNull();
    });

    it(`should reset controls within statement when Product list option is selected
      and also default the addition of UN number`, () => {
      fixture.detectChanges();

      component.substanceReferenceSelect.setValue(PRODUCT_LIST);
      expect(component.tankStatement.get('statement').value).toBeNull();
      expect(component.productListUnNo.controls.length).toBe(1);
    });

    it('should hide section when tankValidationState$ value is HIDDEN', () => {
      spyOn(validationMapper, 'getCurrentState').and.callFake(() =>
        of({ tankDetailsEdit: STATUS.HIDDEN })
      );

      fixture.detectChanges();

      let tankState: STATUS;
      component.tankValidationState$.subscribe(
        ({ tankDetailsEdit }) => (tankState = tankDetailsEdit)
      );

      expect(tankState).toEqual(STATUS.HIDDEN);
      expect(fixture).toMatchSnapshot();
      expect(component.tankDetails.get('tankManufacturer').value).toBeNull();
      expect(component.tankDetails.get('yearOfManufacture').value).toBeNull();
      expect(component.tankDetails.get('tankManufacturerSerialNo').value).toBeNull();
      expect(component.tankDetails.get('tankTypeAppNo').value).toBeNull();
      expect(component.tankDetails.get('tankCode').value).toBeNull();
      expect(component.tankStatement.get('substancesPermitted').value).toBeNull();
    });

    it('should show section when tankValidationState$ value is MANDATORY and required validators are set', () => {
      spyOn(validationMapper, 'getCurrentState').and.callFake(() =>
        of({ tankDetailsEdit: STATUS.MANDATORY })
      );

      fixture.detectChanges();

      let tankState: STATUS;
      component.tankValidationState$.subscribe(
        ({ tankDetailsEdit }) => (tankState = tankDetailsEdit)
      );

      expect(tankState).toEqual(STATUS.MANDATORY);
      expect(component.showTankDetails).toBeTruthy();
      expect(component.tankDetails.get('tankManufacturer').validator).toBeDefined();
      expect(component.tankDetails.get('yearOfManufacture').validator).toBeDefined();
      expect(component.tankDetails.get('tankManufacturerSerialNo').validator).toBeDefined();
      expect(component.tankDetails.get('tankTypeAppNo').validator).toBeDefined();
      expect(component.tankDetails.get('tankCode').validator).toBeDefined();
      expect(component.tankStatement.get('substancesPermitted').validator).toBeDefined();
    });
  });
});
