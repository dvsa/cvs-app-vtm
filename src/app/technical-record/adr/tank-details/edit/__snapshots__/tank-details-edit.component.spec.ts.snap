// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TankDetailsEditComponent handleFormChanges should hide section when tankValidationState$ value is HIDDEN 1`] = `
<vtm-tank-details-edit
  adrForm={[Function FormGroup]}
  booleanOptions={[Function Object]}
  classNumberSubstance={[Function String]}
  detectChange={[Function ViewRef_]}
  fb={[Function FormBuilder]}
  onDestroy$={[Function Subject]}
  parent={[Function FormGroupDirective]}
  permittedSubstance={[Function String]}
  showTankDetails="false"
  substanceProductSelect={[Function String]}
  substanceStatementSelect={[Function String]}
  tank={[Function Object]}
  tankValidationState$={[Function Observable]}
  validationMapper={[Function ValidationMapper]}
>
  
</vtm-tank-details-edit>
`;

exports[`TankDetailsEditComponent should create with initialized form controls 1`] = `
<vtm-tank-details-edit
  adrForm={[Function FormGroup]}
  booleanOptions={[Function Object]}
  classNumberSubstance={[Function String]}
  detectChange={[Function ViewRef_]}
  fb={[Function FormBuilder]}
  onDestroy$={[Function Subject]}
  parent={[Function FormGroupDirective]}
  permittedSubstance={[Function String]}
  showTankDetails={[Function Boolean]}
  substanceProductSelect={[Function String]}
  substanceStatementSelect={[Function String]}
  tank={[Function Object]}
  tankValidationState$={[Function Observable]}
  validationMapper={[Function ValidationMapper]}
>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-three-quarters"
    >
      <div
        class="govuk-form-group"
      >
        <fieldset
          class="govuk-fieldset"
        >
          <form
            class="ng-untouched ng-pristine ng-valid"
            ng-reflect-form="[object Object]"
            novalidate=""
          >
            <h2
              class="govuk-heading-m"
            >
              Tank details
            </h2>
            
            <div
              class="govuk-form-group"
            >
              <label
                class="govuk-label govuk-!-font-weight-bold"
                for="tankManufacturer"
              >
                 Make 
              </label>
              <input
                class="govuk-input govuk-input--width-20 ng-untouched ng-pristine ng-valid"
                formcontrolname="tankManufacturer"
                id="test-edit-tankManufacturer"
                name="tankManufacturer"
                ng-reflect-name="tankManufacturer"
                type="text"
              />
            </div>
            <div
              class="govuk-form-group"
            >
              <label
                class="govuk-label govuk-!-font-weight-bold"
                for="vehicleType"
              >
                 Year of manufacture 
              </label>
              <input
                class="govuk-input govuk-input--width-4 ng-untouched ng-pristine ng-valid"
                formcontrolname="yearOfManufacture"
                id="test-edit-yearOfManufacture"
                name="yearOfManufacture"
                ng-reflect-name="yearOfManufacture"
                ng-reflect-pattern="[0-9]*"
                pattern="[0-9]*"
                type="number"
              />
            </div>
            <div
              class="govuk-form-group"
            >
              <label
                class="govuk-label govuk-!-font-weight-bold"
                for="tankManufacturerSerialNo"
              >
                 Manufacturer serial number 
              </label>
              <input
                class="govuk-input govuk-input--width-10 ng-untouched ng-pristine ng-valid"
                formcontrolname="tankManufacturerSerialNo"
                id="test-edit-tankManufacturerSerialNo"
                name="tankManufacturerSerialNo"
                ng-reflect-name="tankManufacturerSerialNo"
                type="text"
              />
            </div>
            <div
              class="govuk-form-group"
            >
              <label
                class="govuk-label govuk-!-font-weight-bold"
                for="tankTypeAppNo"
              >
                 Tank type approval number 
              </label>
              <input
                class="govuk-input govuk-input--width-10 ng-untouched ng-pristine ng-valid"
                formcontrolname="tankTypeAppNo"
                id="test-edit-tankTypeAppNo"
                name="tankTypeAppNo"
                ng-reflect-name="tankTypeAppNo"
                type="text"
              />
            </div>
            <div
              class="govuk-form-group"
            >
              <label
                class="govuk-label govuk-!-font-weight-bold"
                for="tankCode"
              >
                 Code 
              </label>
              <input
                class="govuk-input govuk-input--width-10 ng-untouched ng-pristine ng-valid"
                formcontrolname="tankCode"
                id="test-edit-tankCode"
                name="tankCode"
                ng-reflect-name="tankCode"
                type="text"
              />
            </div>
            
            <label
              class="govuk-label govuk-!-font-weight-bold"
            >
              Substances permitted
            </label>
            <div
              class="govuk-radios govuk-radios--conditional"
            >
              <div
                class="govuk-radios__item"
              >
                <input
                  class="govuk-radios__input ng-untouched ng-pristine ng-valid"
                  formcontrolname="substancesPermitted"
                  id="test-edit-substancesPermitted-permitted"
                  name="substancesPermitted"
                  ng-reflect-form-control-name="substancesPermitted"
                  ng-reflect-name="substancesPermitted"
                  ng-reflect-value="permitted"
                  type="radio"
                />
                <label
                  class="govuk-label govuk-radios__label"
                  for="test-edit-substancesPermitted-permitted"
                >
                   Substances permitted under the tank code and any special provisions specified in 9 may be carried 
                </label>
              </div>
              <div
                class="govuk-radios__item"
              >
                <input
                  class="govuk-radios__input ng-untouched ng-pristine ng-valid"
                  formcontrolname="substancesPermitted"
                  id="test-edit-substancesPermitted_classNumber"
                  name="substancesPermitted"
                  ng-reflect-form-control-name="substancesPermitted"
                  ng-reflect-name="substancesPermitted"
                  ng-reflect-value="classnumber"
                  type="radio"
                />
                <label
                  class="govuk-label govuk-radios__label"
                  for="test-edit-substancesPermitted_classNumber"
                >
                   Substances (class UN number and if necessary packing group and proper shipping name) may be carried 
                </label>
              </div>
              <div
                class="govuk-inset-text govuk-!-margin-top-0 govuk-radios__conditional--hidden"
                id="test-edit-conditional-substancesPermitted_classNumber"
              >
                
                <div
                  class="govuk-form-group govuk-!-margin-left-3"
                >
                  <label
                    class="govuk-label govuk-!-font-weight-bold"
                    for="substancesPermitted_CLASSNUMBER"
                  >
                     Select: 
                  </label>
                  <div
                    class="govuk-radios"
                  >
                    <div
                      class="govuk-radios__item"
                    >
                      <input
                        class="govuk-radios__input ng-untouched ng-pristine ng-valid"
                        formcontrolname="substanceReferenceSelect"
                        id="test-edit-select-statement-Y"
                        ng-reflect-form-control-name="substanceReferenceSelect"
                        ng-reflect-name="substanceReferenceSelect"
                        ng-reflect-value="statement"
                        type="radio"
                      />
                      <label
                        class="govuk-label govuk-radios__label"
                      >
                         Statement (optional) 
                      </label>
                    </div>
                    <div
                      class="govuk-inset-text govuk-!-margin-top-0 govuk-radios__conditional--hidden"
                      id="conditional-selectStatement"
                    >
                      <div
                        class="govuk-form-group govuk-!-margin-left-3"
                      >
                        <label
                          class="govuk-label govuk-!-font-weight-bold"
                          for="statementRefNumber"
                        >
                           Reference number (optional) 
                        </label>
                        <input
                          class="govuk-input govuk-input--width-10 ng-untouched ng-pristine ng-valid"
                          formcontrolname="statement"
                          id="test-edit-statementRefNumber"
                          name="statementRefNumber"
                          ng-reflect-name="statement"
                          type="text"
                        />
                      </div>
                    </div>
                    <div
                      class="govuk-radios__item"
                    >
                      <input
                        class="govuk-radios__input ng-untouched ng-pristine ng-valid"
                        formcontrolname="substanceReferenceSelect"
                        id="test-edit-select-productlist-N"
                        ng-reflect-form-control-name="substanceReferenceSelect"
                        ng-reflect-name="substanceReferenceSelect"
                        ng-reflect-value="productList"
                        type="radio"
                      />
                      <label
                        class="govuk-label govuk-radios__label"
                      >
                         Product list (optional) 
                      </label>
                    </div>
                    <div
                      class="govuk-inset-text govuk-!-margin-top-0 govuk-radios__conditional--hidden"
                      id="test-edit-conditional-select-productlist-N"
                    >
                      <div
                        class="govuk-form-group govuk-!-margin-left-3"
                      >
                        <div
                          class="govuk-form-group"
                        >
                          <label
                            class="govuk-label govuk-!-font-weight-bold"
                            for="productListRefNumber"
                          >
                             Reference number (optional) 
                          </label>
                          <input
                            class="govuk-input govuk-input--width-10 ng-untouched ng-pristine ng-valid"
                            formcontrolname="productListRefNo"
                            id="test-edit-productListRefNumber"
                            name="productListRefNumber"
                            ng-reflect-name="productListRefNo"
                            type="text"
                          />
                        </div>
                        <div
                          class="govuk-form-group ng-untouched ng-pristine ng-valid"
                          formarrayname="productListUnNo"
                          ng-reflect-name="productListUnNo"
                        >
                          
                          
                          <div
                            class="govuk-form-group"
                          >
                            
                            <label
                              class="govuk-label govuk-!-font-weight-bold"
                              for="productListUnNumber_0"
                            >
                               UN number (optional) 
                            </label>
                            <input
                              class="govuk-input govuk-input--width-5 ng-untouched ng-pristine ng-valid"
                              id="test-edit-productListUnNumber-0"
                              name="productListUnNumber_0"
                              ng-reflect-form="[object Object]"
                              type="text"
                            />
                            
                          </div>
                        </div>
                        <div
                          class="govuk-label govuk-!-margin-top-0 govuk-!-margin-bottom-5"
                        >
                          <a
                            class="govuk-link govuk-link--no-visited-state"
                            href="javascript:void(0)"
                          >
                             Add a UN number 
                          </a>
                        </div>
                        <div
                          class="govuk-form-group"
                        >
                          <label
                            class="govuk-label govuk-!-font-weight-bold"
                            for="productList"
                          >
                             Additional details (optional) 
                          </label>
                          <textarea
                            aria-describedby="productList"
                            class="govuk-textarea ng-untouched ng-pristine ng-valid"
                            formcontrolname="productList"
                            id="test-edit-productList"
                            name="productList"
                            ng-reflect-name="productList"
                            rows="8"
                          >
                                                    
                          </textarea>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            
            
            <div
              class="govuk-form-group"
            >
              <label
                class="govuk-label govuk-!-font-weight-bold"
                for="specialProvisions"
              >
                 Special provisions (optional) 
              </label>
              <textarea
                aria-describedby="specialProvisions"
                class="govuk-textarea govuk-!-margin-bottom-0 ng-untouched ng-pristine ng-valid"
                formcontrolname="specialProvisions"
                id="test-edit-specialProvisions"
                name="specialProvisions"
                ng-reflect-name="specialProvisions"
                rows="8"
              >
                              
              </textarea>
            </div>
          </form>
        </fieldset>
      </div>
    </div>
  </div>
</vtm-tank-details-edit>
`;

exports[`TankDetailsEditComponent should create with multiple UN number controls 1`] = `
<vtm-tank-details-edit
  adrForm={[Function FormGroup]}
  booleanOptions={[Function Object]}
  classNumberSubstance={[Function String]}
  detectChange={[Function ViewRef_]}
  fb={[Function FormBuilder]}
  onDestroy$={[Function Subject]}
  parent={[Function FormGroupDirective]}
  permittedSubstance={[Function String]}
  showTankDetails={[Function Boolean]}
  substanceProductSelect={[Function String]}
  substanceStatementSelect={[Function String]}
  tank={[Function Object]}
  tankValidationState$={[Function Observable]}
  validationMapper={[Function ValidationMapper]}
>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-three-quarters"
    >
      <div
        class="govuk-form-group"
      >
        <fieldset
          class="govuk-fieldset"
        >
          <form
            class="ng-untouched ng-pristine ng-valid"
            ng-reflect-form="[object Object]"
            novalidate=""
          >
            <h2
              class="govuk-heading-m"
            >
              Tank details
            </h2>
            
            <div
              class="govuk-form-group"
            >
              <label
                class="govuk-label govuk-!-font-weight-bold"
                for="tankManufacturer"
              >
                 Make 
              </label>
              <input
                class="govuk-input govuk-input--width-20 ng-untouched ng-pristine ng-valid"
                formcontrolname="tankManufacturer"
                id="test-edit-tankManufacturer"
                name="tankManufacturer"
                ng-reflect-name="tankManufacturer"
                type="text"
              />
            </div>
            <div
              class="govuk-form-group"
            >
              <label
                class="govuk-label govuk-!-font-weight-bold"
                for="vehicleType"
              >
                 Year of manufacture 
              </label>
              <input
                class="govuk-input govuk-input--width-4 ng-untouched ng-pristine ng-valid"
                formcontrolname="yearOfManufacture"
                id="test-edit-yearOfManufacture"
                name="yearOfManufacture"
                ng-reflect-name="yearOfManufacture"
                ng-reflect-pattern="[0-9]*"
                pattern="[0-9]*"
                type="number"
              />
            </div>
            <div
              class="govuk-form-group"
            >
              <label
                class="govuk-label govuk-!-font-weight-bold"
                for="tankManufacturerSerialNo"
              >
                 Manufacturer serial number 
              </label>
              <input
                class="govuk-input govuk-input--width-10 ng-untouched ng-pristine ng-valid"
                formcontrolname="tankManufacturerSerialNo"
                id="test-edit-tankManufacturerSerialNo"
                name="tankManufacturerSerialNo"
                ng-reflect-name="tankManufacturerSerialNo"
                type="text"
              />
            </div>
            <div
              class="govuk-form-group"
            >
              <label
                class="govuk-label govuk-!-font-weight-bold"
                for="tankTypeAppNo"
              >
                 Tank type approval number 
              </label>
              <input
                class="govuk-input govuk-input--width-10 ng-untouched ng-pristine ng-valid"
                formcontrolname="tankTypeAppNo"
                id="test-edit-tankTypeAppNo"
                name="tankTypeAppNo"
                ng-reflect-name="tankTypeAppNo"
                type="text"
              />
            </div>
            <div
              class="govuk-form-group"
            >
              <label
                class="govuk-label govuk-!-font-weight-bold"
                for="tankCode"
              >
                 Code 
              </label>
              <input
                class="govuk-input govuk-input--width-10 ng-untouched ng-pristine ng-valid"
                formcontrolname="tankCode"
                id="test-edit-tankCode"
                name="tankCode"
                ng-reflect-name="tankCode"
                type="text"
              />
            </div>
            
            <label
              class="govuk-label govuk-!-font-weight-bold"
            >
              Substances permitted
            </label>
            <div
              class="govuk-radios govuk-radios--conditional"
            >
              <div
                class="govuk-radios__item"
              >
                <input
                  class="govuk-radios__input ng-untouched ng-pristine ng-valid"
                  formcontrolname="substancesPermitted"
                  id="test-edit-substancesPermitted-permitted"
                  name="substancesPermitted"
                  ng-reflect-form-control-name="substancesPermitted"
                  ng-reflect-name="substancesPermitted"
                  ng-reflect-value="permitted"
                  type="radio"
                />
                <label
                  class="govuk-label govuk-radios__label"
                  for="test-edit-substancesPermitted-permitted"
                >
                   Substances permitted under the tank code and any special provisions specified in 9 may be carried 
                </label>
              </div>
              <div
                class="govuk-radios__item"
              >
                <input
                  class="govuk-radios__input ng-untouched ng-pristine ng-valid"
                  formcontrolname="substancesPermitted"
                  id="test-edit-substancesPermitted_classNumber"
                  name="substancesPermitted"
                  ng-reflect-form-control-name="substancesPermitted"
                  ng-reflect-name="substancesPermitted"
                  ng-reflect-value="classnumber"
                  type="radio"
                />
                <label
                  class="govuk-label govuk-radios__label"
                  for="test-edit-substancesPermitted_classNumber"
                >
                   Substances (class UN number and if necessary packing group and proper shipping name) may be carried 
                </label>
              </div>
              <div
                class="govuk-inset-text govuk-!-margin-top-0 govuk-radios__conditional--hidden"
                id="test-edit-conditional-substancesPermitted_classNumber"
              >
                
                <div
                  class="govuk-form-group govuk-!-margin-left-3"
                >
                  <label
                    class="govuk-label govuk-!-font-weight-bold"
                    for="substancesPermitted_CLASSNUMBER"
                  >
                     Select: 
                  </label>
                  <div
                    class="govuk-radios"
                  >
                    <div
                      class="govuk-radios__item"
                    >
                      <input
                        class="govuk-radios__input ng-untouched ng-pristine ng-valid"
                        formcontrolname="substanceReferenceSelect"
                        id="test-edit-select-statement-Y"
                        ng-reflect-form-control-name="substanceReferenceSelect"
                        ng-reflect-name="substanceReferenceSelect"
                        ng-reflect-value="statement"
                        type="radio"
                      />
                      <label
                        class="govuk-label govuk-radios__label"
                      >
                         Statement (optional) 
                      </label>
                    </div>
                    <div
                      class="govuk-inset-text govuk-!-margin-top-0 govuk-radios__conditional--hidden"
                      id="conditional-selectStatement"
                    >
                      <div
                        class="govuk-form-group govuk-!-margin-left-3"
                      >
                        <label
                          class="govuk-label govuk-!-font-weight-bold"
                          for="statementRefNumber"
                        >
                           Reference number (optional) 
                        </label>
                        <input
                          class="govuk-input govuk-input--width-10 ng-untouched ng-pristine ng-valid"
                          formcontrolname="statement"
                          id="test-edit-statementRefNumber"
                          name="statementRefNumber"
                          ng-reflect-name="statement"
                          type="text"
                        />
                      </div>
                    </div>
                    <div
                      class="govuk-radios__item"
                    >
                      <input
                        class="govuk-radios__input ng-untouched ng-pristine ng-valid"
                        formcontrolname="substanceReferenceSelect"
                        id="test-edit-select-productlist-N"
                        ng-reflect-form-control-name="substanceReferenceSelect"
                        ng-reflect-name="substanceReferenceSelect"
                        ng-reflect-value="productList"
                        type="radio"
                      />
                      <label
                        class="govuk-label govuk-radios__label"
                      >
                         Product list (optional) 
                      </label>
                    </div>
                    <div
                      class="govuk-inset-text govuk-!-margin-top-0 govuk-radios__conditional--hidden"
                      id="test-edit-conditional-select-productlist-N"
                    >
                      <div
                        class="govuk-form-group govuk-!-margin-left-3"
                      >
                        <div
                          class="govuk-form-group"
                        >
                          <label
                            class="govuk-label govuk-!-font-weight-bold"
                            for="productListRefNumber"
                          >
                             Reference number (optional) 
                          </label>
                          <input
                            class="govuk-input govuk-input--width-10 ng-untouched ng-pristine ng-valid"
                            formcontrolname="productListRefNo"
                            id="test-edit-productListRefNumber"
                            name="productListRefNumber"
                            ng-reflect-name="productListRefNo"
                            type="text"
                          />
                        </div>
                        <div
                          class="govuk-form-group ng-untouched ng-pristine ng-valid"
                          formarrayname="productListUnNo"
                          ng-reflect-name="productListUnNo"
                        >
                          
                          
                          <div
                            class="govuk-form-group"
                          >
                            
                            <label
                              class="govuk-label govuk-!-font-weight-bold"
                              for="productListUnNumber_0"
                            >
                               UN number (optional) 
                            </label>
                            <input
                              class="govuk-input govuk-input--width-5 ng-untouched ng-pristine ng-valid"
                              id="test-edit-productListUnNumber-0"
                              name="productListUnNumber_0"
                              ng-reflect-form="[object Object]"
                              type="text"
                            />
                            
                          </div>
                          
                          <div
                            class="govuk-form-group"
                          >
                            
                            <input
                              class="govuk-input govuk-input--width-5 ng-untouched ng-pristine ng-valid"
                              id="test-edit-productListUnNumber-1"
                              name="productListUnNumber_1"
                              ng-reflect-form="[object Object]"
                              type="text"
                            />
                            
                            <span
                              class="govuk-!-font-weight-bold govuk-!-margin-left-2"
                              id="test-edit-removeUNNumber_1"
                            >
                              <a
                                class="govuk-link govuk-link--no-visited-state"
                                href="javascript:void(0)"
                              >
                                <span>
                                  Remove
                                </span>
                              </a>
                            </span>
                          </div>
                        </div>
                        <div
                          class="govuk-label govuk-!-margin-top-0 govuk-!-margin-bottom-5"
                        >
                          <a
                            class="govuk-link govuk-link--no-visited-state"
                            href="javascript:void(0)"
                          >
                             Add a UN number 
                          </a>
                        </div>
                        <div
                          class="govuk-form-group"
                        >
                          <label
                            class="govuk-label govuk-!-font-weight-bold"
                            for="productList"
                          >
                             Additional details (optional) 
                          </label>
                          <textarea
                            aria-describedby="productList"
                            class="govuk-textarea ng-untouched ng-pristine ng-valid"
                            formcontrolname="productList"
                            id="test-edit-productList"
                            name="productList"
                            ng-reflect-name="productList"
                            rows="8"
                          >
                                                    
                          </textarea>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            
            
            <div
              class="govuk-form-group"
            >
              <label
                class="govuk-label govuk-!-font-weight-bold"
                for="specialProvisions"
              >
                 Special provisions (optional) 
              </label>
              <textarea
                aria-describedby="specialProvisions"
                class="govuk-textarea govuk-!-margin-bottom-0 ng-untouched ng-pristine ng-valid"
                formcontrolname="specialProvisions"
                id="test-edit-specialProvisions"
                name="specialProvisions"
                ng-reflect-name="specialProvisions"
                rows="8"
              >
                              
              </textarea>
            </div>
          </form>
        </fieldset>
      </div>
    </div>
  </div>
</vtm-tank-details-edit>
`;
